general/forall.hpp

```diff
@@ -42,6 +42,12 @@ const int MAX_Q1D = 14;
 #define MFEM_UNROLL(N)
 #endif
 
+#if defined(HYPRE_USING_CUDA)
+#define GPU_WRAP1D CuWrap1D
+#elif defined(HYPRE_USING_HIP)
+#define GPU_WRAP1D HipWrap1D
+#endif
```

Comment:
I was thinking of making this macro local to the `.cpp` file where it is needed. If we're going to make it a public macro, it is best to scope it with `MFEM_` prefix and give it a name more similar to the other `MFEM_FORALL` macros -- e.g. something like `MFEM_DEVICE_FORALL` with a definition based on `MFEM_USE_CUDA` and `MFEM_USE_HIP` instead of the HYPRE macros (those, if enabled, need to match the MFEM ones.)

@camierjs, @YohannDudouit , what do you think about this name (`MFEM_DEVICE_FORALL`) for a shorthand name for `CuWrap1D`/`HipWrap1D`?  Or maybe `MFEM_GPU_FORALL` is more appropriate since this is really GPU specific and it is a compile-time "dispatch" unlike `MFEM_FORALL`?

Reply:
I don't have a strong opinion, but I'm not sure we should expose this if this is the only place where it is needed, if we can foresee many other use cases then it may be good to expose. I have a slight preference for `MFEM_DEVICE_FORALL` for consistency reasons.

Reply:
Okay, just to confirm. Since this comes up in two parts of the code, complex_fem.cpp and hypre.cpp, I should define the macro in those two files? The macro being `MFEM_DEVICE_FORALL`?

Reply:
Since it is used in two different `.cpp` files, it is probably better to keep the definition in one place and `forall.hpp` is the right place for that, I think. Regarding the macro name: let's wait to hear from @camierjs as well. Personally, I have a slight preference for `MFEM_GPU_FORALL` which emphasizes that the macro is not meant for MFEM devices like "omp".

Reply:
`MFEM_GPU_FORALL` sounds good to me.
However we are losing the `1D` information if it's mapped to`*Wrap1D`. We might not need the other dimension extensions now, but could in the future.

Reply:
I'm fine with either `MFEM_GPU_FORALL` or `MFEM_DEVICE_FORALL`.

Reply:
I also have a use for this type of functionality, so I think it makes sense to expose it more generally.

In particular, I have a use for something like `MFEM_HYPRE_FORALL`, which would use `MFEM_GPU_FORALL` whenever `HYPRE_USING_GPU` is set, and would fall back on a plain for-loop when it is not. This is useful for writing kernels that interact directly with memory from hypre that is allocated using `HYPRE_MEMORY_DEVICE`.

Reply:
I added `MFEM_GPU__FORALL` and made it consistent with the other macros, I liked @pazner's idea but I would like to propose revisiting outside of this PR.